### PR TITLE
Rearrange the test matrix

### PR DESCRIPTION
**Pull request by RoyCai7**
Optimize test matrix to improve test efficiency
progress ticket: https://progress.opensuse.org/issues/132503

- Related ticket: https://progress.opensuse.org/issues/132530
- Needles: N/A
- Verification run:

XEN tests:
[XEN 12-sp4](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:hansolo:12-SP4:test:&distri=sle&version=12-SP4&groupid=161)
[XEN 12-sp5](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:hansolo:12-SP5:test:&distri=sle&version=12-SP5&groupid=160)
[XEN 15-sp1](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:javor:15-SP1:test:&distri=sle&version=15-SP1&groupid=158)
[XEN 15-sp2](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:yoda:15-SP2:test:&distri=sle&version=15-SP2&groupid=157)
[XEN 15-sp3](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:witch:15-SP3:test:&distri=sle&version=15-SP3&groupid=156)
[XEN 15-sp4](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:hansolo:15-SP4:test:&distri=sle&version=15-SP4&groupid=163)
[XEN 15-sp5](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:xen:darthvader:15-SP5:test:&distri=sle&version=15-SP5&groupid=168)

KVM tests
[KVM 12-sp3](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:jarjarbinks:12-SP3:test:&distri=sle&version=12-SP3&groupid=143)
[KVM 12-sp4](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:darthvader:12-SP4:test:&distri=sle&version=12-SP4&groupid=142)
[KVM 12-sp5](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:darthvader:12-SP5:test:&distri=sle&version=12-SP5&groupid=149)
[KVM 15-sp1](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:javor:15-SP1:test:&distri=sle&version=15-SP1&groupid=147)
[KVM 15-sp2](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:witch:15-SP3:test:&distri=sle&version=15-SP3&groupid=154)
[KVM 15-sp3](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:inara:15-SP2:test:&distri=sle&version=15-SP2&groupid=152)
[KVM 15-sp4](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:hansolo:15-SP4:test:&distri=sle&version=15-SP4&groupid=164)
[KVM 15-sp5](http://openqa.qam.suse.cz/tests/overview?build=:S:M:29711:302738:river:15-SP5:test:&distri=sle&version=15-SP5&groupid=167)

Hyper-V test
[Hyper-v test on 15-sp5](https://openqa.suse.de/tests/11572089)

VMware test
[VMware test on 15-sp5](https://openqa.suse.de/tests/11572090)

### CODE REVIEW

**Comment by RoyCai7** [comment]
Refer matrix:
https://confluence.suse.com/display/Virtualization/Maintenance+Matrix+testing+prio

**Review by tonyyuan1 — APPROVED** [review]

**Comment by tonyyuan1** [comment]
LGTM

**Comment by RoyCai7** [comment]
Keep SLES12SP4 for a period of time though it is LTSS ends.

File: lib/virt_autotest/common.pm [hunk 1]
```diff
@@ -162,10 +162,45 @@ if (get_var("REGRESSION", '') =~ /xen/) {
...
-    delete($guests{sles12sp5HVM}) if (!is_sle('=12-SP5'));
-    delete($guests{sles15sp1HVM}) if (!is_sle('=15-SP1'));
+    # Filter out guests not allowed for the detected SLE version
+    if (is_sle('=12-SP5')) {
+        my @allowed_guests = qw(sles12sp5HVM sles12sp5PV sles15sp1HVM sles15sp1PV sles15sp5HVM sles15sp5PV);
+        foreach my $guest (keys %guests) {
+            delete $guests{$guest} unless grep { $_ eq $guest } @allowed_guests;
+        }
+    } elsif (is_sle('=12-SP4')) {
+        my @allowed_guests = qw(les12sp4HVM sles12sp4PV les12sp5HVM sles12sp5PV);
```
Comment (alice-suse) [inline lib/virt_autotest/common.pm:172]: Typo? les12sp4HVM -> sles12sp4HVM,  les12sp5HVM->sles12sp5HVM?

Comment (RoyCai7) [reply lib/virt_autotest/common.pm]: yes, you are right, will update it.

Comment (RoyCai7) [reply lib/virt_autotest/common.pm]: VR test also does not have these two VM:  sles12sp4HVM and sles12sp5HVM
http://openqa.qam.suse.cz/tests/56924#step/prepare_guests/26

Comment (RoyCai7) [reply lib/virt_autotest/common.pm]: Rerun VR test after fixing.

**Review by alice-suse — CHANGES_REQUESTED** [review]
Hi Roy,
It is amazing to see the PR so soon. Well done :)

From code style, it mainly uses kind of fixed binding for what guests should be kept for which host. It definitely works and simple/clear. That's the absolutely pros. 
- I am not sure if you notice some rule/pattern of it (For host sleAspB, guests sleAspB, sleAspB+1 are mostly needed(only the first and last product are not this way); main exception cases are sle12sp5 host/guest related, because it is the last service pack for sle12, so it is important, and required as guest on the latest two sle15spX host, and requires latest sle15spX guest on its host )? If using the rule/pattern to assign values to @allowed_guests, you can save so many if branch for each product?
- When the next new product sle15spN+1 comes, have you thought about what changes will happen on the required list with above rule/pattern? How to write this PR to save most changing effort?

File: lib/virt_autotest/common.pm (hunk 1 above)
Comment (RoyCai7) [reply lib/virt_autotest/common.pm]: VR Test passed, include sles12sp4HVM sles12sp4PV sles12sp5HVM sles12sp5PV
http://openqa.qam.suse.cz/tests/56930#step/prepare_guests/26

**Comment by RoyCai7** [comment]
> Hi Roy, It is amazing to see the PR so soon. Well done :)
> 
> From code style, it mainly uses kind of fixed binding for what guests should be kept for which host. It definitely works and simple/clear. That's the absolutely pros.
> 
> * I am not sure if you notice some rule/pattern of it (For host sleAspB, guests sleAspB, sleAspB+1 are mostly needed(only the first and last product are not this way); main exception cases are sle12sp5 host/guest related, because it is the last service pack for sle12, so it is important, and required as guest on the latest two sle15spX host, and requires latest sle15spX guest on its host )? If using the rule/pattern to assign values to @allowed_guests, you can save so many if branch for each product?
> * When the next new product sle15spN+1 comes, have you thought about what changes will happen on the required list with above rule/pattern? How to write this PR to save most changing effort?

1. I think the test rules are not very obvious, and may change at any time. Once changed, the changes are uncontrollable. Therefore, the current easy-to-read and easy-to-understand method is used to deal with.
2. When a new product arrives, or the old product is no longer supported, just submit a new PR to add or delete.

Just my opinions.

**Comment by alice-suse** [comment]
> 1. I think the test rules are not very obvious, and may change at any time. Once changed, the changes are uncontrollable. Therefore, the current easy-to-read and easy-to-understand method is used to deal with.
> 2. When a new product arrives, or the old product is no longer supported, just submit a new PR to add or delete.
> 
> Just my opinions.

I understand that the rules may not be that obvious if you are not familiar with how VT project&product managers prioritize things ;-). I can tell because similar rules apply in each new developing product level test plan, sle12sp5(last SP for sle12) and latest sle15 SP are treated with priority so differently.

That's right, the rule may change, but chance is indeed low, based on my experience with them. And even if it changes, new rule/pattern will be present, definitely not random :).

The main pros for controlling guest list with rules/patterns is that we can save changes on this part by someway  when new products come -- only adding the new guest definition, no need to update list for each host.

Well, I do not insist on this . Maybe someday when you guys are bored with repeated changes for each new product, you will think about changing the way. 

Let's also see opinions from @Julie-CAO   @tbaev .

**Comment by Julie-CAO** [comment]
> Well, I do not insist on this . Maybe someday when you guys are bored with repeated changes for each new product, you will think about changing the way.
> 
> Let's also see opinions from @Julie-CAO @tbaev .

both ways have advantages. If ALP would be our next product. I would prefer to enum the list of guests for each host.

**Comment by alice-suse** [comment]
> > 1. I think the test rules are not very obvious, and may change at any time. Once changed, the changes are uncontrollable. Therefore, the current easy-to-read and easy-to-understand method is used to deal with.
> > 2. When a new product arrives, or the old product is no longer supported, just submit a new PR to add or delete.
> > 
> > Just my opinions.
> 
> I understand that the rules may not be that obvious if you are not familiar with how VT project&product managers prioritize things ;-). I can tell because similar rules apply in each new developing product level test plan, sle12sp5(last SP for sle12) and latest sle15 SP are treated with priority so differently.
> 
> That's right, the rule may change, but chance is indeed low, based on my experience with them. And even if it changes, new rule/pattern will be present, definitely not random :).
> 
> The main pros for controlling guest list with rules/patterns is that we can save changes on this part by someway when new products come -- only adding the new guest definition, no need to update list for each host.
> 
> Well, I do not insist on this . Maybe someday when you guys are bored with repeated changes for each new product, you will think about changing the way.
> 
> Let's also see opinions from @Julie-CAO @tbaev .

@tonyyuan1  @tbaev  Hello guys, how do you think? BTW, please be aware that,  when enabling future new products (15sp6 is the next), the guest list per product will need to be checked and some need to be updated with the "hidden" rule in that matrix page. This will be the MAIN difference compared to previous new product enablement.

**Comment by alice-suse** [comment]
https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17391#discussion_r1263327998 is resolved.

**Review by alice-suse — COMMENTED** [review]
The required function is fulfilled and the only issue is fixed. Fine to be merged if most votes for the static guest list definition per host.

**Review by alice-suse — APPROVED** [review]

**Comment by tbaev** [comment]
It looks good to me, good job Roy.

**Comment by alice-suse** [comment]
Roy, you can make it live on void. Congratulations!

**Pull request merged**